File: leetcode/leetcode_cpp/ransom-note.cpp
/*
https://leetcode.com/problems/ransom-note

solution1. map[char] count
time: o(n)
space: o(n)

*/

class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        std::unordered_map<char, int> m;
        for (auto c : magazine) {
            m[c]++;
        }

        for (auto c : ransomNote) {
            if (m[c] == 0)
                return false;
            m[c]--;
        }

        return true;
    }
};

File: leetcode/leetcode_py/ransom-note.py
'''
https://leetcode.com/problems/ransom-note
'''

class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        m = {}
        for c in magazine:
            if c in m:
                m[c] += 1
            else:
                m[c] = 1

        for c in ransomNote:
            if not c in m or m[c] == 0:
                return False
            m[c] -= 1

        return True

        